### PR TITLE
[Feature] Streaming video input with EVS frame filtering (RFC #2201 Phase 2-4)

### PR DESCRIPTION
**Pull request by Sy0307**
## Summary

**Phase 2~4 ready for review. Phase 5 WIP.**

WebSocket-based streaming video input for Qwen3-Omni (RFC #2201 Phase 2–4),
with wire-level async-chunk audio streaming.

- `serving_video_stream.py` — `OmniStreamingVideoHandler` for
  `/v1/video/chat/stream`. Dual-coroutine session loop (`_reader` +
  `_processor` bridged by a bounded `asyncio.Queue`). Consumes
  `OmniRequestOutput` directly from `engine_client.generate()` and
  streams per-step audio as `response.audio.delta`.
- `video_frame_filter.py` — EVS pixel-level pre-filter (64×64 thumbnail
  MSE, default 0.95). ~95% drop on static scenes.
- `video_stream_session.py` — older SSE-based handler kept in tree;
  new `OmniStreamingVideoHandler` supersedes it on the async-chunk path.
- Env switches: `VLLM_VIDEO_ASYNC_CHUNK=on|off` (wire-level streaming
  kill switch, `on` default), `VLLM_VIDEO_AUDIO_DELTA_MODE=fast|slow`
  (incremental vs concat-every-call audio encoding).

### Protocol

```
Client -> Server:
  session.config -> (video.frame | audio.chunk)* -> video.query -> ... -> video.done

Server -> Client:
  response.start -> response.text.delta* -> response.text.done
                 -> response.audio.delta* -> response.audio.done
                 -> (next turn) ... -> session.done
```

### Recent updates (this force-push)

- Rebased onto current `main` so the branch no longer overlaps with
  code already merged via #2581.
- Reviewer feedback addressed in `serving_video_stream.py`:
  only abort prev request on true interrupt (fixes flash_attn
  `scheduler_metadata` crash); bounded audio buffer (4 MB) and
  message queue (200); prewarm tasks tracked and cancelled on
  session exit; `prewarmed_frames` passed explicitly instead of
  stashed on `self`; `_process_query` wrapped in `try/finally`.

### Known issues / scope notes

- **Session KV reuse (Phase 5) is NOT in this PR.** Every `video.query`
  rebuilds the prompt from scratch.
- **Back-to-back short-reply race.** `sessions=3 × turns=4` with several
  `audio_chunks=1-3` turns hits a client-side `Idle timeout` ~1/3 of
  the time. New `add_request` is accepted but stage-0 never emits
  `first_output_latency` while the previous request's trailing outputs
  are still in flight. Engine-layer scheduler race, not addressed here.
  Client workaround: ≥200 ms idle between turns.
- Unit tests currently target the older `VideoStreamHandler`; new
  handler tests are follow-up.

### Dependencies

- Route registration in `api_server.py` depends on #2202 landing.

## Test plan

- [ ] Existing `tests/entrypoints/openai_api/test_video_*` pass.
- [ ] WebSocket end-to-end on 4×H20 with Qwen3-Omni-30B-A3B-Instruct:
      text + audio, multi-turn, EVS on/off,
      `VLLM_VIDEO_ASYNC_CHUNK=on/off`.

### CODE REVIEW

**Comment by Sy0307** [comment]
PTAK @lishunyang12

**Comment by Sy0307** [comment]
This PR will contain #2202 's commit later. Thanks for @lishunyang12 's nice work.

**Comment by Gaohan123** [comment]
Hello, any updates?

**Comment by Sy0307** [comment]
I will carry on this work now.

**Comment by lishunyang12** [comment]
@Sy0307  Any progress? This maybe an important selling point for our 0.20.0 release. Would appreicate if we are closing to merge state.

**Comment by Sy0307** [comment]
@lishunyang12 I have updated this PR and PTAK. Thx.

File: vllm_omni/entrypoints/openai/serving_video_stream.py [hunk 1]
```diff
@@ -0,0 +1,1056 @@
+"""WebSocket handler for streaming video input understanding.
```
Comment (lishunyang12) [inline vllm_omni/entrypoints/openai/serving_video_stream.py:3]: Missing SPDX header — pre-commit will fail. Every other new file in this PR has one.

```suggestion
# SPDX-License-Identifier: Apache-2.0
# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
"""WebSocket handler for streaming video input understanding.
```

File: vllm_omni/entrypoints/openai/serving_video_stream.py [hunk 2]
```diff
@@ -0,0 +1,1056 @@
...
+import torch
+from fastapi import WebSocket, WebSocketDisconnect
+from PIL import Image
+from pydantic import BaseModel, Field, ValidationError
+from vllm.logger import init_logger
+
+from vllm_omni.entrypoints.openai.video_frame_filter import FrameSimilarityFilter
+from vllm_omni.entrypoints.openai.video_stream_context import (
+    text_only_message,
+)
```
Comment (lishunyang12) [inline vllm_omni/entrypoints/openai/serving_video_stream.py:47]: `video_stream_context` isn't added by this PR. Does it land via #2202? Won't import otherwise.

File: tests/entrypoints/openai_api/test_video_stream_handler.py [hunk 3]
```diff
@@ -0,0 +1,593 @@
...
+import io
+import json
+from unittest.mock import AsyncMock
+
+import pytest
+from PIL import Image
+
+from vllm_omni.entrypoints.openai.video_stream_session import (
+    VideoStreamHandler,
+)
```
Comment (lishunyang12) [inline tests/entrypoints/openai_api/test_video_stream_handler.py:23]: These target the old `VideoStreamHandler`; `OmniStreamingVideoHandler` (1056 LOC) has zero runtime tests. Please port before merge.

File: vllm_omni/entrypoints/openai/serving_video_stream.py [hunk 4]
```diff
@@ -0,0 +1,1056 @@
...
+
+# Audio delta extraction mode.
+# - "fast" (default): emit only the new chunk(s) appended since last call.
+#   Matches the "list-of-new-chunks" semantics produced by output_processor
+#   (see output_processor.py:112-123 where audio is appended per-step and
+#   never concatenated). Keeps D2H transfer size independent of answer length.
+# - "slow": pre-fix behaviour. Concatenate the full audio history on every
+#   call, move it all to CPU, then slice. Kept for A/B verification and
+#   scheduled for removal in a future release.
+_AUDIO_DELTA_MODE = os.environ.get("VLLM_VIDEO_AUDIO_DELTA_MODE", "fast").lower()
```
Comment (lishunyang12) [inline vllm_omni/entrypoints/openai/serving_video_stream.py]: Route through `vllm.envs` (or add to `vllm_omni/envs.py`). Reading `os.environ` at import freezes the value, breaks test overrides, and hides these flags from the envs registry. Same for `_ASYNC_CHUNK_MODE` on line 84.

File: vllm_omni/entrypoints/openai/serving_video_stream.py [hunk 5]
```diff
@@ -0,0 +1,1056 @@
...
+                    if msg_type == "video.frame":
+                        frame_data = msg.get("data", "")
+                        if not frame_data:
+                            continue
+                        if len(frame_data) > _MAX_FRAME_SIZE:
+                            await self._send_error(websocket, "Frame too large")
+                            continue
+                        try:
+                            raw_bytes = base64.b64decode(frame_data)
+                            Image.open(io.BytesIO(raw_bytes)).verify()
```
Comment (lishunyang12) [inline vllm_omni/entrypoints/openai/serving_video_stream.py]: `.verify()` here runs on the event loop, then the prewarm task on lines 262–272 decodes again in a thread. Collapse to one `asyncio.to_thread` decode — validation comes free from it raising.

File: vllm_omni/entrypoints/openai/serving_video_stream.py [hunk 6]
```diff
@@ -0,0 +1,1056 @@
...
+                        except json.JSONDecodeError:
+                            continue
+                        choices = data.get("choices", [])
+                        if not choices:
+                            continue
+                        delta = choices[0].get("delta", {})
+                        content = delta.get("content")
+                        if content and isinstance(content, str):
+                            # async_chunk mode splices audio WAV b64 onto text content.
+                            wav_idx = content.find(_WAV_B64_PREFIX)
```
Comment (lishunyang12) [inline vllm_omni/entrypoints/openai/serving_video_stream.py]: Detecting audio via `content.find("UklGR")` inside the text stream splits the reply at the wrong place if that substring ever appears in text. Can we drop `_process_query_chat` or move to a structured delta?

**Review by lishunyang12 — COMMENTED** [review]
User-facing feature — needs a doc. Please add `docs/serving/video_stream_api.md` (mirror the structure of [`docs/serving/speech_api.md`](https://github.com/vllm-project/vllm-omni/blob/main/docs/serving/speech_api.md)) and register it in [`docs/.nav.yml`](https://github.com/vllm-project/vllm-omni/blob/main/docs/.nav.yml) under `User Guide → Serving → OpenAI-Compatible API`. Cover: protocol table (from the file header), `session.config` fields (from `StreamingVideoSessionConfig`), the two env vars, EVS threshold semantics, and the known limitations from the PR body.

**Comment by lishunyang12** [comment]
@linyueqian PTAL

**Comment by Sy0307** [comment]
## Review feedback 

All review suggestions have been solved. See details below. @lishunyang12 

### Code
- New `vllm_omni/entrypoints/openai/video_stream_envs.py`: vLLM-style
  lazy env registry (`environment_variables` + `__getattr__`) for
  `VLLM_VIDEO_ASYNC_CHUNK` / `VLLM_VIDEO_AUDIO_DELTA_MODE`. Handler
  reads them at runtime (no more import-time freeze). Values are
  `.strip().lower()`-normalized; invalid values warn once per
  `(name, value)` pair.
- `serving_video_stream.py`:
  - SPDX header added.
  - Removed event-loop `Image.verify()`. Background prewarm task now
    tags bad frames with a `_BAD_FRAME` sentinel and posts an internal
    `_internal.frame_decode_failed` event; the processor removes the
    frame and sends a `Frame decode failed` error. Client-sent
    `_internal.*` messages are rejected.
  - Removed `_process_query_chat` fallback (including the `UklGR`
    audio-in-text parsing). `chat_service` is retained — it backs
    `_preprocess_to_engine_prompt`'s renderer / chat template.
  - Explicit `ChatCompletionRequest` defaults
    (`add_generation_prompt=True`, `continue_final_message=False`,
    `add_special_tokens=False`).
  - `FrameSimilarityFilter` decode errors now surface as
    `Invalid image data`.
  - `hashlib.md5(..., usedforsecurity=False)` for mm-cache dedupe.
  - `modalities` uses `default_factory`.

### Docs
- New `docs/serving/video_stream_api.md` with protocol tables,
  `session.config` schema (with value ranges), legacy aliases, env
  vars, EVS semantics, and known limitations (including the audio
  overflow clear behaviour and the ≥200 ms idle workaround).
- Registered under **User Guide → Serving → OpenAI-Compatible API** in
  `docs/.nav.yml`.

### Tests
Total: 42 passed (remote vLLM 0.19). New cases:
- `test_video_stream_envs_strip_and_warn_once_per_invalid_value`
- `test_async_chunk_mode_is_read_by_engine_path_at_runtime`
- `test_query_without_engine_client_sends_error`
- `test_frame_prewarm_does_not_block_following_query`
- `test_failed_frame_prewarm_removes_frame_before_query`
- `test_client_cannot_send_internal_frame_decode_failed_message` (new)
- `test_frame_filter_error_sends_invalid_image` (new)
- `test_audio_buffer_overflow_clears_buffer_before_query`
- `test_build_messages_keeps_recent_history_text_only`
- `test_audio_in_video_disabled_omits_mm_processor_kwargs`
- `test_query_inline_audio_data_sets_mm_processor_kwargs`
- `test_video_done_waits_for_in_flight_query`

**Comment by hsliuustc0106** [comment]
it seems https://buildkite.com/vllm/vllm-omni/builds/7906/steps/canvas?sid=019dbeec-6858-4c65-8750-4758df21dbaa&tab=output timeout error not related to this PR, @qibaoyuan PTAL

**Comment by Sy0307** [comment]
Can we merge it? @lishunyang12 @hsliuustc0106
